### PR TITLE
feat(service): Initial user-facing http service 

### PR DESCRIPTION
**Pull request by asonnino**
* [x] Simple http service (minimal) for users
* [x] Simple http client for test
* [x] Make generic storage node to test http service
* [x] Ensure clean shutdown of service
* [x] Improve test coverage

We currently use the simplest API routes, we can discuss the best ones in https://github.com/MystenLabs/walrus/issues/101

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,11 @@
 [workspace]
 resolver = "2"
 members = ["crates/*"]
-default-members = ["crates/walrus-core", "crates/walrus-service", "crates/walrus-test-utils"]
+default-members = [
+    "crates/walrus-core",
+    "crates/walrus-service",
+    "crates/walrus-test-utils",
+]
```
Comment (asonnino) [inline Cargo.toml]: Is my pre-commit misconfigured?

File: crates/walrus-core/src/messages/storage_confirmation.rs [hunk 2]
```diff
@@ -27,6 +31,20 @@ pub struct SignedStorageConfirmation {
     pub signature: BLS12381Signature,
 }
 
+impl SignedStorageConfirmation {
+    /// Creates a new arbitrary signed storage confirmation for tests.
+    pub fn arbitrary_for_test() -> Self {
+        let confirmation = vec![0; 32];
+        let mut rng = StdRng::seed_from_u64(0);
```
Comment (asonnino) [inline crates/walrus-core/src/messages/storage_confirmation.rs]: In the past we had global test-utils generating test keys (that we would re-use in all these types of functions). Should we do this or do we keep generating keys on the spot like here?

File: Cargo.toml (hunk 1 above)
Comment (mlegner) [reply Cargo.toml]: I don't think this is related to pre-commit. Maybe you have configured some formatter for TOML files in your editor?

I see below that you somehow ended up with a corrupted `deny.toml`, maybe that's related? 🤔

Comment (mlegner) [reply Cargo.toml]: In any case, I like this particular change as it makes future additions/removals in this list easier to track with Git.

File: crates/walrus-core/src/lib.rs [hunk 3]
```diff
@@ -48,6 +48,11 @@ impl BlobId {
 
         Self(hasher.finalize().into())
     }
+
+    /// Returns an arbitrary blob ID for testing.
+    pub fn arbitrary_for_test() -> Self {
+        Self([0; Self::LENGTH])
+    }
```
Comment (mlegner) [inline crates/walrus-core/src/lib.rs]: Suggestion: Add a `#[cfg(test)]` in front if we only need it for tests. (Also for other similar functions.)
```suggestion

    /// Returns an arbitrary blob ID for testing.
    #[cfg(test)]
    pub fn arbitrary_for_test() -> Self {
        Self([0; Self::LENGTH])
    }
```

Also: A blob ID of `0` doesn't look very arbitrary to me... 😅

File: crates/walrus-core/src/lib.rs [hunk 4]
```diff
@@ -92,6 +97,11 @@ impl Sliver {
             Sliver::Secondary(_) => SliverType::Secondary,
         }
     }
+
+    /// Returns an arbitrary sliver for testing.
+    pub fn arbitrary_for_test() -> Self {
```
Comment (mlegner) [inline crates/walrus-core/src/lib.rs]: As above:
```suggestion
    #[cfg(test)]
    pub fn arbitrary_for_test() -> Self {
```

File: crates/walrus-core/src/messages/storage_confirmation.rs (hunk 2 above)
Comment (mlegner) [reply crates/walrus-core/src/messages/storage_confirmation.rs]: I think having such functionality that we use in multiple different tests in `walrus-test-utils` would make sense. Then we don't need this boilerplate everywhere. Also, then we can keep `rand` under dev dependencies.

File: crates/walrus-service/src/server.rs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    config::StorageNodePrivateParameters,
+    node::{ServiceState, StoreMetadataError},
+};
+
+/// The path to the store metadata endpoint.
+pub const POST_STORE_BLOB_METADATA: &str = "/storage/metadata";
+/// The path to the store sliver endpoint.
+pub const POST_STORE_SLIVER: &str = "/storage/sliver";
+/// The path to the storage confirmation endpoint.
+pub const GET_STORAGE_CONFIRMATION: &str = "/storage/confirmation";
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs:30]: Suggestion: 
- `/v1/blobs/:blobId/metadata` with GET and PUT
- `/v1/blobs/:blobId/confirmation` with GET
- `/v1/blobs/:blobId/slivers/:sliverPairId/primary` and `/../secondary` with GET and PUT. I would use sliverPairId here, which, along with the blob ID, can be used to determine if and which shard should store it. The shard index could also be used, but the former communicates what is being stored/retrieved as opposed to with whom.

Question: Did you have in mind additional paths beyond `/storage`? I suspect most other services will be on different ports completely.

File: crates/walrus-service/src/server.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+/// Represents a storage sliver request.
+#[derive(Serialize, Deserialize)]
+pub struct StoreSliverRequest {
+    /// The shard index.
+    shard: ShardIndex,
+    /// The blob identifier.
+    blob_id: BlobId,
+    /// The sliver to store.
+    sliver: Sliver,
+}
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs:85]: Suggestion: With the above API endpoints, we can just directly provide the `Sliver` as `application/octet-stream` or `application/json`.

File: crates/walrus-service/src/server.rs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        self,
+        storage_node_private_parameters: &StorageNodePrivateParameters,
+    ) -> JoinHandle<Result<(), std::io::Error>> {
+        let app = Router::new()
+            .route(POST_STORE_BLOB_METADATA, post(Self::store_blob_metadata))
+            .with_state(self.state.clone())
+            .route(POST_STORE_SLIVER, post(Self::store_sliver))
+            .with_state(self.state.clone())
+            .route(GET_STORAGE_CONFIRMATION, get(Self::storage_confirmation))
+            .with_state(self.state.clone());
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs:110]: Suggestion: Let's already add layers for tracing to the router, [for example](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.layer) from `tower_http`.

File: crates/walrus-service/src/server.rs [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+        let mut network_address = storage_node_private_parameters.network_address;
+        network_address.set_ip(IpAddr::V4(Ipv4Addr::UNSPECIFIED));
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: Accept the address (as opposed to the config object) as a parameter.

File: crates/walrus-service/src/server.rs [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+        tokio::spawn(async move {
+            let listener = tokio::net::TcpListener::bind(network_address).await?;
+            axum::serve(listener, app)
+                .with_graceful_shutdown(async move {
+                    let _ = self.shutdown_signal.await;
+                })
+                .await
+        })
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: Accept `self` above as `&self`, then remove the spawn and instead move the body into run(..). This will cause run(..) to block but will give the caller the choice as to whether they want to run the server within their current task or spawn a new one; as opposed to always spawning a new task.

File: crates/walrus-service/src/server.rs [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+        Json(request): Json<BlobMetadataWithId>,
+    ) -> (StatusCode, Result<(), String>) {
+        match state.store_blob_metadata(request) {
+            Ok(()) => (StatusCode::OK, Ok(())),
+            Err(StoreMetadataError::InvalidMetadata(message)) => {
+                (StatusCode::BAD_REQUEST, Err(message.to_string()))
+            }
+            Err(StoreMetadataError::Internal(message)) => {
+                (StatusCode::INTERNAL_SERVER_ERROR, Err(message.to_string()))
+            }
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: For anything we return as `INTERNAL_SERVER_ERROR`, do not provide the message as this can have security implications. Log it instead.

File: crates/walrus-service/src/server.rs [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+            if blob_id.0[0] == 0 {
+                let confirmation = SignedStorageConfirmation::arbitrary_for_test();
+                Ok(Some(StorageConfirmation::Signed(confirmation)))
+            } else if blob_id.0[0] == 1 {
+                Ok(None)
+            } else {
+                Err(anyhow::anyhow!("Invalid shard"))
+            }
+        }
+    }
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: Let's not use a Fake here, we can construct StorageNode instances and use them directly in the test.

File: crates/walrus-service/src/server.rs [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+                    let _ = self.shutdown_signal.await;
+                })
+                .await
+        })
+    }
+
+    async fn store_blob_metadata(
+        State(state): State<Arc<S>>,
+        Json(request): Json<BlobMetadataWithId>,
+    ) -> (StatusCode, Result<(), String>) {
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: Use a consistent JSON body for the errors (for example, [here](https://developers.facebook.com/docs/graph-api/guides/error-handling/) and [here](https://cloud.google.com/storage/docs/json_api/v1/status-codes#404-not-found)). For our purposes, the following should be sufficient for now:
```json
{
  "error": {
    "code": 400,
    "message": "<either message.to_string() or a generic message depending on the error>",   
  }
}
```

File: crates/walrus-service/src/server.rs (hunk 5 above)
Comment (gdanezis) [reply crates/walrus-service/src/server.rs:30]: I strongly support this suggestion to make it easy to add a request router in front of the web servers and shart by blob_id!

File: crates/walrus-core/src/messages/storage_confirmation.rs (hunk 2 above)
Comment (asonnino) [reply crates/walrus-core/src/messages/storage_confirmation.rs]: I agree. On the downside we can't have `walrus-test-utils` under the dev dependencies anymore. Unless we move all test fixtures (such as the function `arbitrary_for_test()`) into the walrus-test-utils crate?

File: crates/walrus-core/src/lib.rs (hunk 3 above)
Comment (asonnino) [reply crates/walrus-core/src/lib.rs]: If we add #[cfg(test)] we can't use this function from other crates it seems. Should we move this
type of functions to the test-utils crate?

File: crates/walrus-core/src/lib.rs (hunk 4 above)
Comment (asonnino) [reply crates/walrus-core/src/lib.rs]: same issue, it seems we can't use the function from other crates if we do that (unless you know a work around?)

File: crates/walrus-service/src/server.rs (hunk 5 above)
Comment (asonnino) [reply crates/walrus-service/src/server.rs:30]: > Question: Did you have in mind additional paths beyond /storage? I suspect most other services will be on different ports completely.

No just used whatever path came to mind with the idea to think them through when fixing https://github.com/MystenLabs/walrus/issues/101

Comment (asonnino) [reply crates/walrus-service/src/server.rs:30]: I updated the https://github.com/MystenLabs/walrus/issues/101 and I can make a fresh PR to improve the paths explicitly (to keep PR small and unblock https://github.com/MystenLabs/walrus/pull/104

File: crates/walrus-service/src/server.rs (hunk 6 above)
Comment (asonnino) [reply crates/walrus-service/src/server.rs:85]: yes this will be much more elegant. added it to the issue and fixing in the next PR

File: crates/walrus-service/src/server.rs (hunk 7 above)
Comment (asonnino) [reply crates/walrus-service/src/server.rs:110]: Should we first land #104 so I can add a unit test for it as well? (opening an issue to not forget, https://github.com/MystenLabs/walrus/issues/106)

File: crates/walrus-core/src/lib.rs (hunk 3 above)
Comment (mlegner) [reply crates/walrus-core/src/lib.rs]: That's true. How about we use the following principle:
- Move test functions that are useful for multiple crates to `walrus-test-utils`.
- Keep more specialized test functions within the corresponding crate, either in the `mod tests` or annotated with `#[cfg(test)]`.

I'm not too happy about having functions for tests lying around everywhere in our normal modules. But that might just be my preference, other people may have different opinions.

Comment (asonnino) [reply crates/walrus-core/src/lib.rs]: I like this approach (this is also what we did in the past and Xun spent a lot of time to do some of it in the Sui codebase)

File: crates/walrus-service/src/server.rs (hunk 12 above)
Comment (mlegner) [reply crates/walrus-service/src/server.rs]: Related to this, although not relevant for devnet: How about we add an [OpenAPI specification](https://spec.openapis.org/oas/latest.html) for our HTTP endpoints? There seems to exist a stable library ([utoipa](https://docs.rs/utoipa/latest/utoipa/)) that helps with this, which is also compatible with axum.

File: crates/walrus-core/src/lib.rs (hunk 3 above)
Comment (mlegner) [reply crates/walrus-core/src/lib.rs]: With "this approach", you refer to having test-util functions located in the standard modules?

As mentioned, I'm not a huge fan of this (as this sometimes has additional dependencies, which then need to be normal instead of dev dependencies), but it's not a hill I'm willing to die on. 🙂

Comment (asonnino) [reply crates/walrus-core/src/lib.rs]: no no I mean the approach you mentioned (adding test-util functions into the test-util module, and keep #[cfg(test)] otherwise)

File: crates/walrus-service/src/server.rs (hunk 11 above)
Comment (asonnino) [reply crates/walrus-service/src/server.rs]: Probably better fixed as a separate PR: https://github.com/MystenLabs/walrus/issues/108

File: deny.toml
Comment (mlegner) [inline deny.toml:1]: Suggestion: Revert the changes to this file and the workspace-level `Cargo.toml`. They don't have any meaningful changes and the reformatting moved some comments to wrong locations.

File: crates/walrus-service/src/node.rs [hunk 13]
```diff
@@ -52,12 +72,11 @@ impl StorageNode {
             n_shards: NonZeroUsize::new(100).unwrap(),
         }
     }
+}
 
+impl ServiceState for StorageNode {
     /// Stores the metadata associated with a blob.
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit: As the functions are already documented in the trait, the docstrings in the implementation can be removed.

File: crates/walrus-service/src/config.rs [hunk 14]
```diff
@@ -65,6 +65,28 @@ pub struct ShardPrivateParameters {
...
+    pub network_address: SocketAddr,
+    /// The metrics address of the shard.
+    pub metrics_address: SocketAddr,
     /// The private parameters shards that the storage node is responsible for.
     pub shards: HashMap<ShardIndex, ShardPrivateParameters>,
 }
+
+impl StorageNodePrivateParameters {
+    /// Creates a new storage node private parameters for testing.
+    pub fn new_for_test() -> Self {
```
Comment (mlegner) [inline crates/walrus-service/src/config.rs]: Nit: If this function is only used for tests within this crate, we can add a `#[cfg(test)]` to it
```suggestion
    /// Creates a new storage node private parameters for testing.
    #[cfg(test)]
    pub fn new_for_test() -> Self {
```

File: Cargo.toml
Comment (mlegner) [inline Cargo.toml:1]: Nit: These changes are not necessary.

File: crates/walrus-service/src/server.rs [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+    async fn store_blob_metadata(
+        State(state): State<Arc<S>>,
+        Json(request): Json<BlobMetadataWithId>,
+    ) -> (StatusCode, Json<ServiceResponse<()>>) {
+        match state.store_blob_metadata(request) {
+            Ok(()) => ServiceResponse::serialized_success(StatusCode::OK, ()),
+            Err(StoreMetadataError::InvalidMetadata(message)) => {
+                ServiceResponse::serialized_error(StatusCode::BAD_REQUEST, message.to_string())
+            }
+            // TODO(@asonnino): Log this error message once #106 lands.
```
Comment (jpcsmith) [inline crates/walrus-service/src/server.rs]: Suggestion: You can already `tracing::error` log these, #106 will just add additional contextual information to the existing log messages.

File: crates/walrus-core/src/metadata.rs [hunk 16]
```diff
@@ -103,6 +103,24 @@ impl UnverifiedBlobMetadataWithId {
             metadata: self.metadata,
         })
     }
+
+    /// Returns an arbitrary metadata object.
+    // todo(@asonnino): Move this function to `walrus-test-utils`.
```
Comment (mlegner) [inline crates/walrus-core/src/metadata.rs]: Open TODO.
